File: .migration-backup-20251210_115339/docs/diagrams/export-to-png.js
#!/usr/bin/env node

/**
 * Export Excalidraw diagram to PNG
 *
 * This script uses Puppeteer to render the Excalidraw diagram in a headless browser
 * and export it as a PNG image at high resolution.
 *
 * Requirements:
 *     npm install puppeteer
 */

const fs = require('fs');
const path = require('path');
const puppeteer = require('puppeteer');

const EXCALIDRAW_FILE = path.join(__dirname, 'jpspec-workflow.excalidraw');
const OUTPUT_PNG = path.join(__dirname, 'jpspec-workflow.png');
const WIDTH = 1920;
const SCALE = 2; // For high DPI

async function exportToPng() {
  console.log('Loading Excalidraw diagram...');
  const excalidrawData = JSON.parse(fs.readFileSync(EXCALIDRAW_FILE, 'utf8'));

  console.log('Launching browser...');
  const browser = await puppeteer.launch({
    headless: 'new',
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  try {
    const page = await browser.newPage();

    // Set viewport to desired width
    await page.setViewport({
      width: WIDTH,
      height: 1200,
      deviceScaleFactor: SCALE
    });

    // Create HTML with embedded Excalidraw viewer
    const html = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <style>
    body {
      margin: 0;
      padding: 20px;
      background: white;
    }
    #canvas {
      width: 100%;
      height: 100%;
    }
  </style>
</head>
<body>
  <canvas id="canvas"></canvas>
  <script type="module">
    const excalidrawData = ${JSON.stringify(excalidrawData)};

    // Simple canvas renderer for Excalidraw elements
    const canvas = document.getElementById('canvas');
    const ctx = canvas.getContext('2d');

    // Calculate bounds
    let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
    excalidrawData.elements.forEach(el => {
      minX = Math.min(minX, el.x);
      minY = Math.min(minY, el.y);
      maxX = Math.max(maxX, el.x + (el.width || 0));
      maxY = Math.max(maxY, el.y + (el.height || 0));
    });

    const padding = 50;
    const width = maxX - minX + padding * 2;
    const height = maxY - minY + padding * 2;

    canvas.width = width * ${SCALE};
    canvas.height = height * ${SCALE};
    canvas.style.width = width + 'px';
    canvas.style.height = height + 'px';

    ctx.scale(${SCALE}, ${SCALE});
    ctx.translate(-minX + padding, -minY + padding);

    // Render elements
    excalidrawData.elements.forEach(el => {
      ctx.save();

      // Set styles
      ctx.strokeStyle = el.strokeColor || '#000';
      ctx.fillStyle = el.backgroundColor || 'transparent';
      ctx.lineWidth = el.strokeWidth || 1;
      ctx.globalAlpha = (el.opacity || 100) / 100;

      if (el.strokeStyle === 'dashed') {
        ctx.setLineDash([5, 5]);
      }

      if (el.type === 'rectangle') {
        const radius = el.roundness?.type === 3 ? 8 : 0;
        if (radius > 0) {
          // Rounded rectangle
          ctx.beginPath();
          ctx.moveTo(el.x + radius, el.y);
          ctx.lineTo(el.x + el.width - radius, el.y);
          ctx.quadraticCurveTo(el.x + el.width, el.y, el.x + el.width, el.y + radius);
          ctx.lineTo(el.x + el.width, el.y + el.height - radius);
          ctx.quadraticCurveTo(el.x + el.width, el.y + el.height, el.x + el.width - radius, el.y + el.height);
          ctx.lineTo(el.x + radius, el.y + el.height);
          ctx.quadraticCurveTo(el.x, el.y + el.height, el.x, el.y + el.height - radius);
          ctx.lineTo(el.x, el.y + radius);
          ctx.quadraticCurveTo(el.x, el.y, el.x + radius, el.y);
          ctx.closePath();
        } else {
          ctx.beginPath();
          ctx.rect(el.x, el.y, el.width, el.height);
          ctx.closePath();
        }

        if (el.backgroundColor !== 'transparent') {
          ctx.fill();
        }
        ctx.stroke();
      } else if (el.type === 'diamond') {
        ctx.beginPath();
        const cx = el.x + el.width / 2;
        const cy = el.y + el.height / 2;
        ctx.moveTo(cx, el.y);
        ctx.lineTo(el.x + el.width, cy);
        ctx.lineTo(cx, el.y + el.height);
        ctx.lineTo(el.x, cy);
        ctx.closePath();

        if (el.backgroundColor !== 'transparent') {
          ctx.fill();
        }
        ctx.stroke();
      } else if (el.type === 'arrow') {
        const points = el.points;
        if (points && points.length > 0) {
          ctx.beginPath();
          ctx.moveTo(el.x + points[0][0], el.y + points[0][1]);
          for (let i = 1; i < points.length; i++) {
            ctx.lineTo(el.x + points[i][0], el.y + points[i][1]);
          }
          ctx.stroke();

          // Draw arrowhead
          if (el.endArrowhead === 'arrow' && points.length > 1) {
            const lastPoint = points[points.length - 1];
            const prevPoint = points[points.length - 2];
            const angle = Math.atan2(lastPoint[1] - prevPoint[1], lastPoint[0] - prevPoint[0]);
            const arrowSize = 10;

            ctx.beginPath();
            ctx.moveTo(el.x + lastPoint[0], el.y + lastPoint[1]);
            ctx.lineTo(
              el.x + lastPoint[0] - arrowSize * Math.cos(angle - Math.PI / 6),
              el.y + lastPoint[1] - arrowSize * Math.sin(angle - Math.PI / 6)
            );
            ctx.moveTo(el.x + lastPoint[0], el.y + lastPoint[1]);
            ctx.lineTo(
              el.x + lastPoint[0] - arrowSize * Math.cos(angle + Math.PI / 6),
              el.y + lastPoint[1] - arrowSize * Math.sin(angle + Math.PI / 6)
            );
            ctx.stroke();
          }
        }
      } else if (el.type === 'text') {
        ctx.font = \`\${el.fontSize || 16}px Arial\`;
        ctx.fillStyle = el.strokeColor || '#000';
        ctx.textAlign = el.textAlign || 'left';
        ctx.textBaseline = el.verticalAlign || 'top';

        const lines = el.text.split('\\n');
        const lineHeight = (el.fontSize || 16) * (el.lineHeight || 1.25);

        lines.forEach((line, i) => {
          let x = el.x;
          if (el.textAlign === 'center') {
            x += el.width / 2;
          } else if (el.textAlign === 'right') {
            x += el.width;
          }

          ctx.fillText(line, x, el.y + i * lineHeight);
        });
      }

      ctx.restore();
    });

    // Signal completion
    window.renderComplete = true;
  </script>
</body>
</html>
    `;

    await page.setContent(html);

    // Wait for render to complete
    await page.waitForFunction(() => window.renderComplete, { timeout: 10000 });

    // Add small delay for final render
    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('Taking screenshot...');
    const canvas = await page.$('#canvas');
    await canvas.screenshot({
      path: OUTPUT_PNG,
      omitBackground: false
    });

    console.log(`✓ PNG exported to: ${OUTPUT_PNG}`);
    console.log(`  Dimensions: ${WIDTH}px width (${SCALE}x scale)`);

  } finally {
    await browser.close();
  }
}

// Run export
exportToPng().catch(err => {
  console.error('Export failed:', err);
  process.exit(1);
});

File: docs/diagrams/export-to-png.py
#!/usr/bin/env python3
"""
Export Excalidraw diagram to PNG

This script renders the Excalidraw JSON diagram to a PNG image
using PIL (Pillow) for high-quality output.

Requirements:
    pip install Pillow
"""

import json
import math
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont

# Configuration
EXCALIDRAW_FILE = Path(__file__).parent / "flowspec-workflow.excalidraw"
OUTPUT_PNG = Path(__file__).parent / "flowspec-workflow.png"
SCALE = 2  # For high DPI
PADDING = 50


def hex_to_rgb(hex_color):
    """Convert hex color to RGB tuple"""
    hex_color = hex_color.lstrip("#")
    return tuple(int(hex_color[i : i + 2], 16) for i in (0, 2, 4))


def draw_rounded_rectangle(draw, x, y, width, height, radius, fill, outline, width_px):
    """Draw a rounded rectangle"""
    # Main rectangles
    draw.rectangle(
        [x + radius, y, x + width - radius, y + height], fill=fill, outline=None
    )
    draw.rectangle(
        [x, y + radius, x + width, y + height - radius], fill=fill, outline=None
    )

    # Corners
    draw.ellipse([x, y, x + radius * 2, y + radius * 2], fill=fill, outline=None)
    draw.ellipse(
        [x + width - radius * 2, y, x + width, y + radius * 2], fill=fill, outline=None
    )
    draw.ellipse(
        [x, y + height - radius * 2, x + radius * 2, y + height],
        fill=fill,
        outline=None,
    )
    draw.ellipse(
        [x + width - radius * 2, y + height - radius * 2, x + width, y + height],
        fill=fill,
        outline=None,
    )

    # Outline
    if outline:
        # Top and bottom
        draw.line([x + radius, y, x + width - radius, y], fill=outline, width=width_px)
        draw.line(
            [x + radius, y + height, x + width - radius, y + height],
            fill=outline,
            width=width_px,
        )

        # Left and right
        draw.line([x, y + radius, x, y + height - radius], fill=outline, width=width_px)
        draw.line(
            [x + width, y + radius, x + width, y + height - radius],
            fill=outline,
            width=width_px,
        )

        # Corner arcs
        draw.arc(
            [x, y, x + radius * 2, y + radius * 2],
            180,
            270,
            fill=outline,
            width=width_px,
        )
        draw.arc(
            [x + width - radius * 2, y, x + width, y + radius * 2],
            270,
            360,
            fill=outline,
            width=width_px,
        )
        draw.arc(
            [x, y + height - radius * 2, x + radius * 2, y + height],
            90,
            180,
            fill=outline,
            width=width_px,
        )
        draw.arc(
            [x + width - radius * 2, y + height - radius * 2, x + width, y + height],
            0,
            90,
            fill=outline,
            width=width_px,
        )


def render_excalidraw_to_png():
    """Render Excalidraw JSON to PNG"""
    print("Loading Excalidraw diagram...")
    with open(EXCALIDRAW_FILE, "r") as f:
        data = json.load(f)

    elements = data["elements"]

    # Calculate bounds
    min_x = min_y = float("inf")
    max_x = max_y = float("-inf")

    for el in elements:
        min_x = min(min_x, el["x"])
        min_y = min(min_y, el["y"])
        max_x = max(max_x, el["x"] + el.get("width", 0))
        max_y = max(max_y, el["y"] + el.get("height", 0))

    # Calculate image dimensions
    width = int((max_x - min_x + PADDING * 2) * SCALE)
    height = int((max_y - min_y + PADDING * 2) * SCALE)

    print(f"Creating image: {width}x{height} px")

    # Create image
    img = Image.new("RGB", (width, height), color="white")
    draw = ImageDraw.Draw(img)

    # Transform coordinates
    def transform_x(x):
        return int((x - min_x + PADDING) * SCALE)

    def transform_y(y):
        return int((y - min_y + PADDING) * SCALE)

    # Load font with cross-platform support
    try:
        font_cache = {}

        def get_font(size):
            size_scaled = int(size * SCALE)
            if size_scaled not in font_cache:
                # Try multiple font paths for cross-platform support
                font_paths = [
                    "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",  # Linux
                    "/Library/Fonts/Arial.ttf",  # macOS
                    "C:\\Windows\\Fonts\\arial.ttf",  # Windows
                    "/usr/share/fonts/truetype/freefont/FreeSans.ttf",  # Linux alternative
                ]
                for font_path in font_paths:
                    try:
                        font_cache[size_scaled] = ImageFont.truetype(
                            font_path, size_scaled
                        )
                        break
                    except Exception:
                        continue
                else:
                    # Fallback to default font if none found
                    font_cache[size_scaled] = ImageFont.load_default()
            return font_cache[size_scaled]
    except Exception:

        def get_font(size):  # noqa: ARG001
            return ImageFont.load_default()

    print("Rendering elements...")

    # Sort elements by type to ensure proper layering (shapes first, then arrows, then text)
    sorted_elements = sorted(
        elements,
        key=lambda el: (
            0
            if el["type"] in ["rectangle", "diamond"]
            else 1
            if el["type"] == "arrow"
            else 2,
            el.get("id", ""),
        ),
    )

    for el in sorted_elements:
        el_type = el["type"]
        x = transform_x(el["x"])
        y = transform_y(el["y"])

        # Parse colors
        stroke_color = el.get("strokeColor", "#1e1e1e")
        bg_color = el.get("backgroundColor", "transparent")
        opacity = el.get("opacity", 100) / 100

        stroke_rgb = hex_to_rgb(stroke_color) if stroke_color != "transparent" else None
        fill_rgb = hex_to_rgb(bg_color) if bg_color != "transparent" else None

        # Apply opacity
        if fill_rgb and opacity < 1:
            fill_rgb = tuple(int(c * opacity + 255 * (1 - opacity)) for c in fill_rgb)

        stroke_width = int(el.get("strokeWidth", 1) * SCALE)

        if el_type == "rectangle":
            w = int(el["width"] * SCALE)
            h = int(el["height"] * SCALE)
            roundness = el.get("roundness", {})

            if roundness.get("type") == 3:
                # Rounded rectangle
                radius = int(8 * SCALE)
                draw_rounded_rectangle(
                    draw, x, y, w, h, radius, fill_rgb, stroke_rgb, stroke_width
                )
            else:
                # Regular rectangle
                if fill_rgb:
                    draw.rectangle([x, y, x + w, y + h], fill=fill_rgb)
                if stroke_rgb:
                    draw.rectangle(
                        [x, y, x + w, y + h], outline=stroke_rgb, width=stroke_width
                    )

        elif el_type == "diamond":
            w = int(el["width"] * SCALE)
            h = int(el["height"] * SCALE)
            cx = x + w // 2
            cy = y + h // 2

            points = [(cx, y), (x + w, cy), (cx, y + h), (x, cy)]

            if fill_rgb:
                draw.polygon(points, fill=fill_rgb)
            if stroke_rgb:
                draw.polygon(points, outline=stroke_rgb, width=stroke_width)

        elif el_type == "arrow":
            points = el.get("points", [])
            if len(points) > 1:
                # Draw line segments
                for i in range(len(points) - 1):
                    x1 = transform_x(el["x"] + points[i][0])
                    y1 = transform_y(el["y"] + points[i][1])
                    x2 = transform_x(el["x"] + points[i + 1][0])
                    y2 = transform_y(el["y"] + points[i + 1][1])

                    if stroke_rgb:
                        draw.line([x1, y1, x2, y2], fill=stroke_rgb, width=stroke_width)

                # Draw arrowhead
                if el.get("endArrowhead") == "arrow" and len(points) > 1:
                    last = points[-1]
                    prev = points[-2]

                    x1 = transform_x(el["x"] + last[0])
                    y1 = transform_y(el["y"] + last[1])

                    angle = math.atan2(last[1] - prev[1], last[0] - prev[0])
                    arrow_size = 10 * SCALE

                    # Left arrow line
                    x2 = x1 - arrow_size * math.cos(angle - math.pi / 6)
                    y2 = y1 - arrow_size * math.sin(angle - math.pi / 6)
                    draw.line([x1, y1, x2, y2], fill=stroke_rgb, width=stroke_width)

                    # Right arrow line
                    x3 = x1 - arrow_size * math.cos(angle + math.pi / 6)
                    y3 = y1 - arrow_size * math.sin(angle + math.pi / 6)
                    draw.line([x1, y1, x3, y3], fill=stroke_rgb, width=stroke_width)

        elif el_type == "text":
            text = el.get("text", "")
            font_size = el.get("fontSize", 16)
            font = get_font(font_size)

            lines = text.split("\n")
            line_height = int(font_size * el.get("lineHeight", 1.25) * SCALE)

            for i, line in enumerate(lines):
                text_y = y + i * line_height

                # Text color
                text_color = hex_to_rgb(stroke_color)

                # Draw text
                draw.text((x, text_y), line, fill=text_color, font=font)

    print(f"Saving PNG to: {OUTPUT_PNG}")
    img.save(OUTPUT_PNG, "PNG", optimize=True)
    print("✓ PNG exported successfully!")
    print(
        f"  Dimensions: {width}x{height} px ({width // SCALE}x{height // SCALE} logical)"
    )


if __name__ == "__main__":
    try:
        render_excalidraw_to_png()
    except Exception as e:
        print(f"Error: {e}")
        import traceback

        traceback.print_exc()
        exit(1)
